src/backend/distributed/connection/placement_connection.c

```diff
@@ -837,7 +837,18 @@ CheckShardPlacements(ConnectionShardHashEntry *shardEntry,
 
 		if (placementEntry->failed)
 		{
-			UpdateShardPlacementState(placementEntry->key.placementId, FILE_INACTIVE);
+			uint64 shardId = shardEntry->key.shardId;
+			uint64 placementId = placementEntry->key.placementId;
+			ShardPlacement *shardPlacement = LoadShardPlacement(shardId, placementId);
+
+			/*
+			 * We only set shard state if its current state is FILE_FINALIZED, which
+			 * prevents overwriting shard state if it is already set at somewhere else.
+			 */
+			if (shardPlacement->shardState == FILE_FINALIZED)
+			{
+				UpdateShardPlacementState(placementEntry->key.placementId, FILE_INACTIVE);
+			}
 		}
```

Comment:
I'd like to see this change in its own commit, with its own justification.  While this seems like a reasonable safety measure, I don't quite see why this is needed for this change?  It's certainly not something that should be subsumed in a commit about DropShards without really being mentioned...